### PR TITLE
[SPARK-11353][IO] Update jets3t version to 0.9.4

### PR DESCRIPTION
**Pull request by lpiepiora**
This PR updates jets3t dependency to 0.9.4, because of an error, which is thrown when code tries to write to S3 bucket located in Frankfurt.

### CODE REVIEW

**Comment by srowen** [comment]
I think this is OK since it's a maintenance release, but are there any other changes across the two releases that might be an issue?

**Comment by markgrover** [comment]
LGTM too. HADOOP-9623 updated the jets3t version to 0.9.0 and it went in Hadoop 2.3.0, so Hadoop starting 2.3.0 or later should be fine. The only noteworthy thing from jets3t release notes was:
`NOTE: Anyone who has implemented their own JetS3t service implemented the JetS3tRequestAuthorizer will need to adjust their code due to API changes.`
A quick grep through spark code revealed no reference to JetS3tRequestAuthorizer so I think we should be ok.

**Comment by lpiepiora** [comment]
Seems that JetS3t bumped their version of HttpClient to 4.5, which is conflicting with HtmlUnit, which actually tries to read private fields of HttpClientBuilder. Since the field name changed (from `sslcontext` to `sslContext`) the patch fails.

I've tested it by building Spark locally, and we should be good by sticking to Spark's provided version of HttpClient, so I'll exclude this (and HttpCore) transient dependency from the jets3t library.

Beside I've looked through Hadoop sources and compared it against [diff between 0.9.3 and 0.9.4] (https://bitbucket.org/jmurty/jets3t/branches/compare/Release-0.9.4%0DRelease-0.9.3#diff). The files affected by the change between 0.9.3 and 0.9.4 doesn't seem to touch any public members used directly by Hadoop code.

**Comment by lpiepiora** [comment]
Jenkins, retest this please

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by srowen** [comment]
@lpiepiora OK I believe Jenkins, that this is a real failure. It is the `UISeleniumSuite`. Something about the output of a UI has changed: possibly an error. Are you able to reproduce that?

**Comment by lpiepiora** [comment]
Yes, thanks - I'm able to reproduce it. Now, I'm trying to figure out what's wrong.

**Comment by steveloughran** [comment]
Moving to Hadoop 0.90 [HADOOP-9623](https://issues.apache.org/jira/browse/HADOOP-9623) was what could be described as "an accidental disaster"'; the patch swallowed exceptions "which should never happen", resulting in [HADOOP-10589](https://issues.apache.org/jira/browse/HADOOP-10589); a seek(0) on a 0-byte file NPE-ing. (trivia: It was fixed by probably the only piece of co-recursive code in core hadoop)

One issue with 0.90 is that the `close()` call on an input stream reads _all remaining bytes on the resource_ [HADOOP-12376](https://issues.apache.org/jira/browse/HADOOP-12376). This hurts: moving up to 0.94 may fix it. From the hadoop core perspective, the move to 0.90 broke enough things that we are scared to go near the s3n code again; all future work is in s3a.

To summarise then: this may break s3n if not shaded, but you should be encouraging people to use s3a on Hadoop 2.7+ anyway

**Comment by lpiepiora** [comment]
@steveloughran yes, that's exactly what happened to me in this PR. I wanted to fix it but, in general as you've said this just yields more problems on multiple levels. 

However s3a is not a breeze either (even in newer Hadoop 2.7+ versions), especially with Frankfurt buckets, which support only AWS Signature V4.

I'll close this PR anyway, because I think this is not the right way either (even though this jets3t update was a minor one it upgraded transitive dependencies, which yielded multiple issues).

**Comment by steveloughran** [comment]
> However s3a is not a breeze either (even in newer Hadoop 2.7+ versions), especially with Frankfurt buckets, which support only AWS Signature V4.

really? thought that worked. I know HADOOP-12537 mentioned it, but didn't think STS credentials were mandatory. As usual: file a JIRA.

**Comment by Kinghack** [comment]
since latest version of jets3t does not build into latest spark. Is it possible for now that I could access s3 file from regions that supports AWS4-HMAC-SHA256 only?
